crates/matrix-sdk-crypto-js/src/machine.rs

```diff
@@ -0,0 +1,549 @@
+    ///
+    /// Sessions need to be established between devices so group
+    /// sessions for a room can be shared with them.
+    ///
+    /// This should be called every time a group session needs to be
+    /// shared as well as between sync calls. After a sync some
+    /// devices may request room keys without us having a valid Olm
+    /// session with them, making it impossible to server the room key
+    /// request, thus it’s necessary to check for missing sessions
+    /// between sync as well.
+    ///
+    /// Note: Care should be taken that only one such request at a
+    /// time is in flight, e.g. using a lock.
+    ///
+    /// The response of a successful key claiming requests needs to be
+    /// passed to the `OlmMachine` with the `mark_request_as_sent`.
+    ///
+    /// `users` represents the list of users that we should check if
+    /// we lack a session with one of their devices. This can be an
+    /// empty iterator when calling this method between sync requests.
+    #[wasm_bindgen(js_name = "getMissingSessions")]
+    pub fn get_missing_sessions(&self, users: &Array) -> Result<Promise, JsError> {
+        let users = users
+            .iter()
+            .map(|user| Ok(downcast::<identifiers::UserId>(&user, "UserId")?.inner.clone()))
+            .collect::<Result<Vec<ruma::OwnedUserId>, JsError>>()?;
+
+        let me = self.inner.clone();
+
+        Ok(future_to_promise(async move {
+            match me.get_missing_sessions(users.iter().map(AsRef::as_ref)).await? {
+                Some((transaction_id, keys_claim_request)) => {
+                    Ok(JsValue::from(requests::KeysClaimRequest::try_from((
+                        transaction_id.to_string(),
+                        &keys_claim_request,
+                    ))?))
+                }
+
+                None => Ok(JsValue::NULL),
+            }
+        }))
+    }
+
+    /// Get a verification object for the given user ID with the given flow ID.
+    ///
+    /// Returns a list of `JsValue` to represent either (depending on
+    /// how the Wasm module has been compiled):
+    ///   * `Sas` (enabled),
+    ///   * `Qr`
+    #[cfg_attr(feature = "qrcode", doc = "(enabled).")]
+    #[cfg_attr(not(feature = "qrcode"), doc = "(disabled).")]
+    ///
+    /// If a verification mode is missing, please try to compile the
+    /// Wasm module with different features.
+    #[wasm_bindgen(js_name = "getVerification")]
+    pub fn get_verification(
+        &self,
+        user_id: &identifiers::UserId,
+        flow_id: &str,
+    ) -> Result<JsValue, JsError> {
+        self.inner
+            .get_verification(user_id.inner.as_ref(), flow_id)
+            .map(verifications::Verification)
+            .map(JsValue::try_from)
+            .transpose()
+            .map(|r| r.unwrap_or(JsValue::UNDEFINED))
+    }
+}
+
+/// An Ed25519 public key, used to verify digital signatures.
+#[wasm_bindgen]
+#[derive(Debug, Clone)]
+pub struct Ed25519PublicKey {
```

Comment:
Not sure if people on the Javascript side want to deal with this type, libolm returns all the public keys as `base64` encoded strings so only the libolm side needs to deal with validation. Vodozemac supports both ways quite well.

Something to discuss with the Javascript folks.

Reply:
We generally don't want to expose the `verify()` method. A valid signature does not necessarily mean that an object is trusted, it just means that the signature if valid for this particular public key.

It's quite complicated to determine if something should be trusted and very easy to get wrong. We do have some higher level methods for certain objects that do tell you if something is trusted or not.

Reply:
Let's create a new issue ticket listing this issue item, as part of the next work to do to finish Wasm+JS.